### DISCUSSION TITLE
Missing modelSlice methods?

### DISCUSSION THREAD

**Discussion by jalaziz** [stephenafamo/bob#310] [Q&A] [answered]
The 0.29.0 changelog states:

> - `table.InsertQ` has been renamed to `table.Insert`. The old implementation of `Insert` has been removed.  
   The same functionality can be achieved by using `modelSlice.Insert()` or creating an `Insert` query using `table.Insert()`.
> - `table.UpdateQ` has been renamed to `table.Update`. The old implementation of `Update` has been removed.  
   The same functionality can be achieved by using `modelSlice.Update()` or creating an `Update` query using `table.Update()`.
> - `table.DeleteQ` has been renamed to `table.Delete`. The old implementation of `Delete` has been removed.  
   The same functionality can be achieved by using `modelSlice.Delete()` or creating an `Delete` query using `table.Delete()`.
  
However, there are no `Insert`, `Update`, or `Delete` methods on the model slice types. The only available methods are `DeleteAll`, `UpdateAll`, `ReloadAll`, `UpdateMod`, `DeleteMod`, and `AfterQueryHook`.

Additionally, the current [model documentation](https://bob.stephenafamo.com/docs/models/table#insert) continues to refer to the old, non-existent methods.

Are these methods missing? Is the changelog mistaken? What is the current recommended pattern for inserting, updating, and deleting using the generated models?

**Comment by stephenafamo** [comment, marked as answer]
The docs definitely need to be updated.   
There have been quite a few large changes recently, and the docs have not been updated to match.

The CHANGELOG also has a mistake....

## Insert

### Old way

Previously you could insert a slice of **Setters** with `table.Insert` and `table.InsertMany`:

```go
user, err := models.Users.Insert(ctx, db, setter) // insert one
users, err := models.Users.InsertMany(ctx, db, setters...) // insert many
```

### New way

Now, each setter is an insert mod by itself:

```go
user, err := models.Users.Insert(setter).One(ctx, db) // insert one
users, err := models.Users.Insert(setters[0], setters[1]).All(ctx, db) // insert many


// For cases where you already have a slice of setters and you want to pass them all, you can use `bob.ToMods`
users, err := models.Users.Insert(bob.ToMods(setters)).All(ctx, db) // insert many
```

## Update and Delete

This one is more a typo, `modelSlice.UpdateAll` and `modelSlice.DeleteAll` are what should have been referenced.

However, the implementation is quite trivial if you take a look at the generated code

```go
func (o UserSlice) UpdateAll(ctx context.Context, exec bob.Executor, vals UserSetter) error {
	_, err := Users.Update(vals.UpdateMod(), o.UpdateMod()).All(ctx, exec)
	return err
}

func (o UserSlice) DeleteAll(ctx context.Context, exec bob.Executor) error {
	_, err := Users.Delete(o.DeleteMod()).Exec(ctx, exec)
	return err
}
```

**Reply by jalaziz** [reply]
If I'm not mistaken, there's still a typo in your example. Should it be?

```
user, err := models.Users.Insert(setter).One(ctx, db) // insert one
users, err := models.Users.Insert(setters[0], setters[1]).All(ctx, db) // insert many

// For cases where you already have a slice of setters and you want to pass them all, you can use `bob.ToMods`
users, err := models.Users.Insert(bob.ToMods(setters)).All(ctx, db) // insert many
```

**Reply by stephenafamo** [reply]
You're correct... fixed
